### DISCUSSION TITLE
Meaning of "importers" in pnpm-lock.yaml?

### DISCUSSION THREAD

**Discussion by MartinX3** [pnpm/pnpm#6214] [Q&A] [answered]
Using renovatebot as the dependency manager it adds always
```yaml
importers:

  .:
    specifiers:
```
at the beginning of my pnpm-lock.yaml

But if I `pnpm install` on my local computer, it removes the
```yaml
importers:

  .:
```
part and only
```yaml
specifiers:
```
is left.

Why does this happen?
What is the meaning of `importers` and `.:`?

PS:
Sometimes it happens in reverse.
So my local computer adds the stuff and the renovate dependency PR removes it again.

**Comment by stefanmo** [comment]
I have the same issue and would like to know. It's driving me insane.

**Comment by gluxon** [comment, marked as answer]
The "_importers_" block refers to each internal package within a monorepo. For example, if you have the following directory structure in a repo:

```
❯ tree .
.
├── package.json
└── packages
    ├── a
    │   └── package.json
    └── b
        └── package.json
```

The lockfile will look like:

```yaml
importers:
  .:
    specifiers:
      # ...

  packages/a:
    specifiers:
      # ...

  packages/b:
    specifiers:
      # ...
```

For repositories with only a single `package.json` (i.e. not a monorepo), pnpm doesn't create the `importers` block by default. It just removes the `importers` block and puts what would have been in `importers["."]` one level higher.

```yaml
specifiers:
  # ...
```

## Flickering

>  So my local computer adds the stuff and the renovate dependency PR removes it again.

From a quick test, pnpm decides which format to use based on the presence of the `pnpm-workspace.yaml` file. Is it possible that file is present locally, but not present in the job renovate runs in? It might be `.gitignore`'ed. Even if the file is empty, pnpm will think you want a monorepo and create an `importers` block.

**Reply by MartinX3** [reply]
Thank you for the response.

This file is not git ignored and i have only a single package.json.

I hope it won't happen again, since I updated to pnpm 8 now.

**Reply by gluxon** [reply]
Interesting. If this does happen on pnpm 8 in the future for you, it'd be good to confirm if `pnpm-workspace.yaml` was present during `pnpm install`. Workspaces can also be [configured in the root `package.json`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#workspaces), so it's possible some tooling is adding setting that field.

It's also possible pnpm just has a bug. If there's a clear repro that would help significantly.

**Reply by MartinX3** [reply]
But you said the workspace space file is only for monorepos
Thanks for the info.
I created this file because a npm migration guide told me to do so in the past.
I'll try to remove it.

(I wish thw github app would allow me to edit my posts)

**Reply by gluxon** [reply]
Yup! If the `pnpm-workspace.yml` file is present, deleting it should prevent pnpm from creating the `importers` block in the future. This file is indeed only for monorepos.

Thanks for checking.

**Reply by MartinX3** [reply]
Thank you, too. :)

So is it possible a bug between the github renovate bot and pnpm if I have a workspace file with the following ocntent in my project?

```yaml
packages:
    # include packages in subfolders (e.g. apps/ and packages/)
    - 'apps/**'
    - 'packages/**'
    # if required, exclude some directories
    - '!**/test/**'
```

**Reply by gluxon** [reply]
Do you have `package.json` files in the `apps` and `packages`? If so, you likely want the `pnpm-workspace.yaml` file and pnpm should be creating the `importers` block in `pnpm-lock.yaml`. It would be strange if renovate creates a pull request without that block.

**Reply by MartinX3** [reply]
No, there are no other package.json files

**Reply by gluxon** [reply]
Cool. I'd recommend deleting `pnpm-workspace.yaml` and you should never have this problem again. 🙂

**Reply by MartinX3** [reply]
It works, thank you. :)

But I wonder why this weird switching in the lock file happened in the first place all the time.
